File: tools/python/scell_context_string_echo.py
#
# Copyright (c) 2017 Cossack Labs Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import sys
from base64 import b64encode, b64decode
from pythemis import scell

_, COMMAND, KEY, MESSAGE, CONTEXT = range(5)

if len(sys.argv) != 5:
    print('Usage: <command: enc | dec > <key> <message> <context>')
    exit(1)

command = sys.argv[COMMAND]
key = sys.argv[KEY]
message = sys.argv[MESSAGE]
context = sys.argv[CONTEXT]
cell = scell.SCellContextImprint(key.encode('utf-8'))
if command == 'enc':
    encrypted = cell.encrypt(message.encode('utf-8'), context.encode('utf-8'))
    encoded = b64encode(encrypted)
    print(encoded.decode('ascii'))
elif command == 'dec':
    decoded = b64decode(message.encode('utf-8'))
    decrypted = cell.decrypt(decoded, context.encode('utf-8'))
    print(decrypted.decode('utf-8'))
else:
    print('Wrong command, use "enc" or "dec"')
    exit(1)

File: tools/rust/scell_context_string_echo.rs
// Copyright 2019 (c) rust-themis developers
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::process::exit;

use clap::clap_app;
use themis::secure_cell::SecureCell;

fn main() {
    let matches = clap_app!(scell_context_string_echo =>
        (version: env!("CARGO_PKG_VERSION"))
        (about: "Secure Cell echo testing tool (context imprint mode).")
        (@arg mode:    +required "<enc|dec>")
        (@arg key:     +required "master key")
        (@arg message: +required "message to encrypt or decrypt")
        (@arg context: +required "user context")
    )
    .get_matches();

    let mode = matches.value_of("mode").unwrap();
    let key = matches.value_of("key").unwrap();
    let message = matches.value_of("message").unwrap();
    let context = matches.value_of("context").unwrap();

    let cell = SecureCell::with_key(&key)
        .unwrap_or_else(|_| {
            eprintln!("invalid parameters: empty master key");
            exit(1);
        })
        .context_imprint();

    match mode {
        "enc" => {
            let encrypted = cell
                .encrypt_with_context(&message, &context)
                .unwrap_or_else(|error| {
                    eprintln!("failed to encrypt message: {}", error);
                    exit(1);
                });
            println!("{}", base64::encode(&encrypted));
        }
        "dec" => {
            let decoded_message = base64::decode(&message).unwrap_or_else(|error| {
                eprintln!("failed to decode message: {}", error);
                exit(1);
            });
            let decrypted = cell
                .decrypt_with_context(&decoded_message, &context)
                .unwrap_or_else(|error| {
                    eprintln!("failed to decrypt message: {}", error);
                    exit(1);
                });
            println!("{}", std::str::from_utf8(&decrypted).expect("UTF-8 string"));
        }
        other => {
            eprintln!("invalid mode {}, use \"enc\" or \"dec\"", other);
            exit(1);
        }
    }
}
